### PR TITLE
WELD-2250 Fix a scenario with qualifier class present in multiple arc…

### PR DESCRIPTION
**Pull request by manovotn**
…hives within EAR. Add automated test.

### CODE REVIEW

File: impl/src/main/java/org/jboss/weld/resolution/QualifierInstance.java [hunk 1]
```diff
@@ -134,7 +136,22 @@ public QualifierInstance(Class<? extends Annotation> annotationClass, Map<String
                     }
                     builder.put(method.getJavaMember().getName(), method.getJavaMember().invoke(instance));
                 } catch (IllegalAccessException | InvocationTargetException | IllegalArgumentException e) {
-                    throw ResolutionLogger.LOG.cannotCreateQualifierInstanceValues(instance, Formats.formatAsStackTraceElement(method.getJavaMember()), e);
+                    // it may happen that we are in EAR and have stored the annotation's method from different WAR class loader
+                    // an invocation will then lead to IAE, we can re-try with reflection
+                    Method[] methods;
+                    builder = ImmutableMap.builder();
+                    if (System.getSecurityManager() != null) {
+                        methods = AccessController.doPrivileged((PrivilegedAction<Method[]>)() -> annotationClass.getDeclaredMethods());
```
Comment (mkouba) [inline impl/src/main/java/org/jboss/weld/resolution/QualifierInstance.java]: You could use `GetDeclaredMethodsAction`.

Comment (manovotn) [reply impl/src/main/java/org/jboss/weld/resolution/QualifierInstance.java]: Hmm, I was looking for one such. But failing to find it I just did this.

File: impl/src/main/java/org/jboss/weld/resolution/QualifierInstance.java [hunk 2]
```diff
@@ -134,7 +136,22 @@ public QualifierInstance(Class<? extends Annotation> annotationClass, Map<String
...
+                    } else {
+                        methods = annotationClass.getDeclaredMethods();
+                    }
+                    for (Method m : methods) {
```
Comment (mkouba) [inline impl/src/main/java/org/jboss/weld/resolution/QualifierInstance.java]: We have to filter out members annotated with `javax.enterprise.util.Nonbinding`.

File: impl/src/main/java/org/jboss/weld/resolution/QualifierInstance.java [hunk 3]
```diff
@@ -134,7 +137,24 @@ public QualifierInstance(Class<? extends Annotation> annotationClass, Map<String
                     }
                     builder.put(method.getJavaMember().getName(), method.getJavaMember().invoke(instance));
                 } catch (IllegalAccessException | InvocationTargetException | IllegalArgumentException e) {
-                    throw ResolutionLogger.LOG.cannotCreateQualifierInstanceValues(instance, Formats.formatAsStackTraceElement(method.getJavaMember()), e);
+                    // it may happen that we are in EAR and have stored the annotation's method from different WAR class loader
```
Comment (mkouba) [inline impl/src/main/java/org/jboss/weld/resolution/QualifierInstance.java]: I think we should only re-try if `IllegalArgumentException` is thrown following the `java.lang.reflect.Method.invoke(Object, Object...)` javadoc.

File: tests-arquillian/src/test/java/org/jboss/weld/tests/qualifier/ear/LibraryWithQualifierInTwoWarsTest.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+        // setup EAR
+        EnterpriseArchive ear = ShrinkWrap.create(EnterpriseArchive.class, Utils.getDeploymentNameAsHash(LibraryWithQualifierInTwoWarsTest.class, Utils.ARCHIVE_TYPE.EAR))
+            .addAsModules(war1, war2);
+        return ear;
+    }
+
+    @Test
+    public void test() {
+        // verification is merely the ability to deploy this scenario
```
Comment (mkouba) [inline tests-arquillian/src/test/java/org/jboss/weld/tests/qualifier/ear/LibraryWithQualifierInTwoWarsTest.java]: Maybe we should doublecheck the relevant qualifier instance is created... but you're hopefully writing a failing test first so it should be ok ;-)

Comment (manovotn) [reply tests-arquillian/src/test/java/org/jboss/weld/tests/qualifier/ear/LibraryWithQualifierInTwoWarsTest.java]: Yes, the test was written prior to the fix, so it was failing.

**Pull request merged**